// ice/src/es5/Ice/WSEndpointFactory.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

//
// Copyright (c) ZeroC, Inc. All rights reserved.
//
var Ice = require("../Ice/WSEndpoint").Ice;

var WSEndpoint = Ice.WSEndpoint;

var WSEndpointFactory =
/*#__PURE__*/
function () {
  function WSEndpointFactory(instance, delegate) {
    _classCallCheck(this, WSEndpointFactory);

    this._instance = instance;
    this._delegate = delegate;
  }

  _createClass(WSEndpointFactory, [{
    key: "type",
    value: function type() {
      return this._instance.type();
    }
  }, {
    key: "protocol",
    value: function protocol() {
      return this._instance.protocol();
    }
  }, {
    key: "create",
    value: function create(args, oaEndpoint) {
      var e = new WSEndpoint(this._instance, this._delegate.create(args, oaEndpoint));
      e.initWithOptions(args);
      return e;
    }
  }, {
    key: "read",
    value: function read(s) {
      var e = new WSEndpoint(this._instance, this._delegate.read(s));
      e.initWithStream(s);
      return e;
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this._delegate.destroy();

      this._instance = null;
    }
  }]);

  return WSEndpointFactory;
}();

Ice.WSEndpointFactory = WSEndpointFactory;
exports.Ice = Ice;
